src/util/ser.rs

```diff
@@ -0,0 +1,344 @@
+				Ok(ret)
+			}
+		}
+	);
+}
+
+impl_array!(32);
+impl_array!(33);
+
+// Tuples
+macro_rules! tuple_encode {
+	($($x:ident),*) => (
+		impl<W: Write, $($x: Writeable<W>),*> Writeable<W> for ($($x),*) {
+			#[inline]
+			#[allow(non_snake_case)]
+			fn write(&self, w: &mut Writer<W>) -> Result<(), DecodeError> {
+					let &($(ref $x),*) = self;
+					$( $x.write(w)?; )*
+					Ok(())
+			}
+		}
+		impl<R: Read, $($x: Readable<R>),*> Readable<R> for ($($x),*) {
+			#[inline]
+			#[allow(non_snake_case)]
+			fn read(r: &mut Reader<R>) -> Result<($($x),*), DecodeError> {
+					Ok(($({let $x = $x::read(r)?; $x }),*))
+			}
+		}
+	);
+}
+
+tuple_encode!(T0, T1);
+tuple_encode!(T0, T1, T2, T3);
+tuple_encode!(T0, T1, T2, T3, T4, T5);
+tuple_encode!(T0, T1, T2, T3, T4, T5, T6, T7);
+
+
+// References
+impl<W: Write, T: Writeable<W>> Writeable<W> for Box<T> {
+	#[inline]
+	fn write(&self, w: &mut Writer<W>) -> Result<(), DecodeError> { (**self).write(w) }
+}
+
+impl<R: Read, T: Readable<R>> Readable<R> for Box<T> {
+	#[inline]
+	fn read(r: &mut Reader<R>) -> Result<Box<T>, DecodeError> {
+		Ok(Box::new(T::read(r)?))
+	}
+}
+
+// HashMap
+impl<W, K, V> Writeable<W> for HashMap<K, V>
+	where W: Write,
+		  K: Writeable<W> + Eq + Hash,
+		  V: Writeable<W>
+{
+	#[inline]
+	fn write(&self, w: &mut Writer<W>) -> Result<(), DecodeError> {
+	(self.len() as u16).write(w)?;
+		for (key, value) in self.iter() {
+			key.write(w)?;
+			value.write(w)?;
+		}
+		Ok(())
+	}
+}
+
+impl<R, K, V> Readable<R> for HashMap<K, V>
+	where R: Read,
+				K: Readable<R> + Eq + Hash,
+				V: Readable<R>
+{
+	#[inline]
+	fn read(r: &mut Reader<R>) -> Result<HashMap<K, V>, DecodeError> {
+		let len: u16 = Readable::read(r)?;
+		let mut ret = HashMap::with_capacity(len as usize);
+		for _ in 0..len {
+				ret.insert(K::read(r)?, V::read(r)?);
+		}
+		Ok(ret)
+	}
+}
+
+impl<W: Write, T: Writeable<W>> Writeable<W> for [T] {
+	#[inline]
+	fn write(&self, w: &mut Writer<W>) -> Result<(), DecodeError> {
+		(self.len() as u16).write(w)?;
+		for c in self.iter() { c.write(w)?; }
+		Ok(())
+	}
+}
+
+// Vectors
+impl<W: Write, T: Writeable<W>> Writeable<W> for Vec<T> {
+	#[inline]
+	fn write(&self, w: &mut Writer<W>) -> Result<(), DecodeError> {
+		let byte_size = (self.len() as usize)
+											.checked_mul(mem::size_of::<T>())
+											.ok_or(DecodeError::InvalidLength)?;
+		if byte_size > MAX_VEC_SIZE {
+				return Err(DecodeError::InvalidLength);
+		}
+		(&self[..]).write(w)
+	}
+}
+
+impl<R: Read, T: Readable<R>> Readable<R> for Vec<T> {
+	#[inline]
+	fn read(r: &mut Reader<R>) -> Result<Vec<T>, DecodeError> {
+			let len: u16 = Readable::read(r)?;
+			let byte_size = (len as usize)
+												.checked_mul(mem::size_of::<T>())
+												.ok_or(DecodeError::InvalidLength)?;
+			if byte_size > MAX_VEC_SIZE {
+					return Err(DecodeError::InvalidLength);
+			}
+			let mut ret = Vec::with_capacity(len as usize);
```

Comment:
You need something to ensure that there is at least len capacity left in the buffer to be read, as allocating full-usize-memory will crash us.

Reply:
I am not sure what you meant. `len*sizeof(T)` is bounded and each `read()?` will raise 'ShortRead' if there is not enough data.

Reply:
Ah, I guess I'm blind. That said, MAX_VEC_SIZE is much too large for network messages (which can never be logner than 16K), even if its within reason for ChannelMonitors :/